sgp30/sgp30.go

```diff
@@ -0,0 +1,193 @@
+
+package sgp30
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"log"
+	"strconv"
+	"time"
+
+	"periph.io/x/conn/v3"
+	"periph.io/x/conn/v3/i2c"
+)
+
+const (
+	initAirQuality       uint16 = 0x2003
+	measureAirQuality    uint16 = 0x2008
+	getIAQBaseline       uint16 = 0x2015
+	setIAQBaseline       uint16 = 0x201e
+	setHumidity          uint16 = 0x2061
+	measureTest          uint16 = 0x2032
+	getFeatureSetVersion uint16 = 0x202f
+	measureRawSignals    uint16 = 0x2050
+	getTVOCBaseline      uint16 = 0x20b3
+	setTVOCBaseline      uint16 = 0x2077
+
+	I2CAddress uint16 = 0x58
+)
+
+// commandDuration maps the defined maximum measurement duration from the sensor
+var commandDuration = map[uint16]time.Duration{
+	initAirQuality:       time.Millisecond * 10,
+	measureAirQuality:    time.Millisecond * 12,
+	getIAQBaseline:       time.Millisecond * 10,
+	setIAQBaseline:       time.Millisecond * 10,
+	setHumidity:          time.Millisecond * 10,
+	measureTest:          time.Millisecond * 220,
+	getFeatureSetVersion: time.Millisecond * 10,
+	measureRawSignals:    time.Millisecond * 25,
+	getTVOCBaseline:      time.Millisecond * 10,
+	setTVOCBaseline:      time.Millisecond * 10,
+}
+
+// commandResponseLength maps the defined response length including the CRC
+var commandResponseLength = map[uint16]int{
+	measureAirQuality:    6,
+	getIAQBaseline:       6,
+	measureTest:          3,
+	getFeatureSetVersion: 3,
+	measureRawSignals:    6,
+	getTVOCBaseline:      3,
+}
+
+// CO2 represents the current carbon dioxide value in ppm
+type CO2 uint16
+
+func (c CO2) String() string {
+	return strconv.Itoa(int(c)) + "ppm"
+}
+
+func (c *CO2) set(b []byte) {
+	*c = (CO2)(binary.BigEndian.Uint16(b))
+}
+
+// TVOC represents the current total volatile organic compounds value in ppb
+type TVOC uint16
+
+func (t TVOC) String() string {
+	return strconv.Itoa(int(t)) + "ppb"
+}
+
+func (t *TVOC) set(b []byte) {
+	*t = (TVOC)(binary.BigEndian.Uint16(b))
+}
+
+// Env represents measurements from an environmental sensor.
+type Env struct {
+	CO2  CO2
+	TVOC TVOC
+}
+
+// NewI2C returns an object that communicates over I2C to SGP30 environmental sensor.
+//
+// The address must be 0x58.
+func NewI2C(b i2c.Bus, addr uint16, ctx context.Context) (*Dev, error) {
+	switch addr {
+	case I2CAddress:
+	default:
+		return nil, errors.New("sgp30: given address not supported by device")
+	}
+	d := &Dev{
+		d: &i2c.Dev{Bus: b, Addr: addr},
+		env: Env{
+			CO2:  400,
+			TVOC: 0,
+		},
+	}
+	if err := d.makeDev(ctx); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// Dev is a handle to an initialized SGP30 device.
+type Dev struct {
+	d   conn.Conn
+	env Env
+}
+
+// EquivalentCO2 returns the latest equivalent CO2 value
+func (d *Dev) EquivalentCO2() CO2 {
+	return d.env.CO2
+}
+
+// TotalVOC returns the latest total VOC value
+func (d *Dev) TotalVOC() TVOC {
+	return d.env.TVOC
+}
+
+func (d *Dev) makeDev(ctx context.Context) error {
```

Comment:
I do plan to change all drivers to use this form. I think I'd prefer it to exported as ReadLoop(ctx) and be blocking. Then it's up to the caller to use a go routine instead of the driver.

Reply:
Do you also plan to add an interface for the devices? Should I implement this now or will you (or should I) adjust this later?

Reply:
It can be done later, it's fine for now.